auth_service/auth_adapter/core/auth.py

```diff
@@ -18,121 +18,207 @@
+    else:
+        # user already exists in the registry
+        context = _compare_user(user, external_claims)
+        if context:
+            user = _get_inactivated_user(user, context)
+            try:
+                await user_dao.update(user)
+            except Exception as exc:  # pylint:disable=broad-except
+                log.warning("Error updating user: %s", exc)
+        internal_claims.update(id=user.id, status=user.status)
+    internal_token = sign_and_encode_token(internal_claims)
     return internal_token
 
 
-def decode_and_verify_token(
+def decode_and_validate_token(
     access_token: Optional[str], key=None
 ) -> Optional[dict[str, Any]]:
-    """Decode and verify the given JSON Web Token."""
+    """Decode and validate the given JSON Web Token."""
     if not access_token:
         return None
     if not key:
-        key = signing_keys.external_jwks
+        key = jwt_config.external_jwks
         if not key:
-            log.debug("No external signing key, cannot verify token")
+            log.debug("No external signing key, cannot validate token.")
             return None
-    jws_token = jws.JWS()
     try:
-        jws_token.deserialize(access_token, key=key)
-        payload = json.loads(jws_token.payload.decode("UTF-8"))
-        return payload
-    except jws.InvalidJWSObject:
-        log.debug("Invalid access token format")
-    except jws.InvalidJWSSignature:
-        log.debug("Invalid access token signature")
-    except jws.JWKeyNotFound:
-        log.debug("Signature key for access token not found")
-    except UnicodeDecodeError:
-        log.debug("Access token payload has invalid encoding")
-    except json.JSONDecodeError:
-        log.debug("Access token payload is not valid JSON")
+        token = jwt.JWT(
+            jwt=access_token,
+            key=key,
+            algs=jwt_config.external_algs,
+            check_claims=jwt_config.check_claims,
+            expected_type="JWS",
+        )
+        claims = json.loads(token.claims)
+        # in addition to JWT.validate() which checks whether claims exist,
+        # we also make sure that the copied claims are not null or empty strings
+        if not claims["sub"]:
+            raise ValueError("The subject claim is missing.")
+        for claim in jwt_config.copied_claims:
+            if not claims[claim]:
+                raise ValueError(f"Missing value for {claim} claim.")
+        return claims
+    except Exception as exc:  # pylint:disable=broad-except
+        log.debug("Cannot validate external token: %s", exc)
     return None
 
 
-def sign_and_encode_token(payload: dict[str, Any], key=None) -> Optional[str]:
+def sign_and_encode_token(claims: dict[str, Any], key=None) -> Optional[str]:
     """Encode and sign the given payload as JSON Web Token.
 
     Returns None in case the payload could not be properly signed.
     """
-    if not payload:
+    if not claims:
         return None
```

Comment:
Maybe again handle the case that there are no claims needed outside of this function? For a function called like this, I am expecting that it either returns me a valid token (str) or fails but I wouldn't expect None.

Reply:
Changed in 2a334db.